{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Cell/Tumor Input Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import pickle\n",
    "\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import typing as t\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from cdrpy.util.io import read_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_folder = Path(\"../../../data/datasets/GDSCv2DepMap\")\n",
    "input_folder = Path(\"../../../data/inputs/GDSCv2DepMap\")\n",
    "genelist_folder =  Path(\"../../../data/genelists\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "screen_data = pd.read_csv(dataset_folder / \"ScreenDoseResponseLabels.csv\")\n",
    "\n",
    "# screen_data[[\"id\", \"cell_id\", \"drug_id\", \"ln_ic50\"]].rename(\n",
    "#     columns={\"ln_ic50\": \"label\"}\n",
    "# ).to_csv(input_folder / \"LabelsLogIC50.csv\", index=False)\n",
    "\n",
    "# screen_data[[\"id\", \"cell_id\", \"drug_id\", \"auc\"]].rename(\n",
    "#     columns={\"auc\": \"label\"}\n",
    "# ).to_csv(input_folder / \"LabelsAUC.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load gene lists\n",
    "mcg_genes = pd.read_csv(genelist_folder / \"VermaatMiniCancerGenome.csv\")\n",
    "mcg_genes = set(mcg_genes[\"HGNC_SYMBOL\"].to_list())\n",
    "\n",
    "cgc_genes = pd.read_csv(genelist_folder / \"CancerGeneCensusFeb062023.csv\")\n",
    "cgc_genes = set(cgc_genes[\"Gene Symbol\"].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dualgcn_folder = Path(\"../../../data/raw/DualGCN\")\n",
    "dualgcn_genes = read_list(dualgcn_folder / \"gene_list.txt\")\n",
    "dualgcn_genes = set(dualgcn_genes)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load omics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SIDM00046</th>\n",
       "      <td>3.636915</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.084064</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>3.049631</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>5.273889</td>\n",
       "      <td>4.166715</td>\n",
       "      <td>0.201634</td>\n",
       "      <td>...</td>\n",
       "      <td>3.958843</td>\n",
       "      <td>6.162693</td>\n",
       "      <td>0.773996</td>\n",
       "      <td>1.731183</td>\n",
       "      <td>4.032982</td>\n",
       "      <td>0.443607</td>\n",
       "      <td>2.684819</td>\n",
       "      <td>6.743623</td>\n",
       "      <td>4.218781</td>\n",
       "      <td>3.653060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00078</th>\n",
       "      <td>5.130107</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>3.759156</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>5.973382</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.499527</td>\n",
       "      <td>4.775051</td>\n",
       "      <td>0.790772</td>\n",
       "      <td>...</td>\n",
       "      <td>3.709291</td>\n",
       "      <td>5.216843</td>\n",
       "      <td>1.443607</td>\n",
       "      <td>3.090853</td>\n",
       "      <td>4.078097</td>\n",
       "      <td>1.422233</td>\n",
       "      <td>2.333424</td>\n",
       "      <td>6.576976</td>\n",
       "      <td>3.486714</td>\n",
       "      <td>2.691534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00079</th>\n",
       "      <td>5.152995</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.328406</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>5.972693</td>\n",
       "      <td>3.885574</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>...</td>\n",
       "      <td>4.898208</td>\n",
       "      <td>6.947082</td>\n",
       "      <td>1.090853</td>\n",
       "      <td>2.157044</td>\n",
       "      <td>4.101818</td>\n",
       "      <td>2.659925</td>\n",
       "      <td>3.275007</td>\n",
       "      <td>6.530757</td>\n",
       "      <td>3.179511</td>\n",
       "      <td>3.477677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00080</th>\n",
       "      <td>4.175525</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>6.417009</td>\n",
       "      <td>0.124328</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.555816</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>5.590961</td>\n",
       "      <td>3.532317</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.373648</td>\n",
       "      <td>6.251151</td>\n",
       "      <td>0.757023</td>\n",
       "      <td>1.803227</td>\n",
       "      <td>3.493135</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>2.495695</td>\n",
       "      <td>5.282440</td>\n",
       "      <td>3.845992</td>\n",
       "      <td>4.080658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00081</th>\n",
       "      <td>4.822730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.120601</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>2.778209</td>\n",
       "      <td>0.176323</td>\n",
       "      <td>4.913129</td>\n",
       "      <td>3.283922</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>...</td>\n",
       "      <td>4.051372</td>\n",
       "      <td>6.105175</td>\n",
       "      <td>0.895303</td>\n",
       "      <td>1.922198</td>\n",
       "      <td>4.047015</td>\n",
       "      <td>0.963474</td>\n",
       "      <td>2.803227</td>\n",
       "      <td>5.291309</td>\n",
       "      <td>3.544733</td>\n",
       "      <td>4.852498</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19193 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A1BG      A1CF       A2M     A2ML1   A3GALT2    A4GALT  \\\n",
       "model_id                                                                \n",
       "SIDM00046  3.636915  0.042644  0.321928  0.084064  0.028569  3.049631   \n",
       "SIDM00078  5.130107  0.014355  0.150560  3.759156  0.056584  5.973382   \n",
       "SIDM00079  5.152995  0.028569  0.111031  0.097611  0.000000  4.328406   \n",
       "SIDM00080  4.175525  0.042644  6.417009  0.124328  0.000000  0.555816   \n",
       "SIDM00081  4.822730  0.000000  6.120601  0.028569  0.042644  2.778209   \n",
       "\n",
       "              A4GNT      AAAS      AACS     AADAC  ...    ZWILCH     ZWINT  \\\n",
       "model_id                                           ...                       \n",
       "SIDM00046  0.028569  5.273889  4.166715  0.201634  ...  3.958843  6.162693   \n",
       "SIDM00078  0.000000  5.499527  4.775051  0.790772  ...  3.709291  5.216843   \n",
       "SIDM00079  0.042644  5.972693  3.885574  0.111031  ...  4.898208  6.947082   \n",
       "SIDM00080  0.097611  5.590961  3.532317  0.000000  ...  4.373648  6.251151   \n",
       "SIDM00081  0.176323  4.913129  3.283922  0.056584  ...  4.051372  6.105175   \n",
       "\n",
       "               ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "model_id                                                                \n",
       "SIDM00046  0.773996  1.731183  4.032982  0.443607  2.684819  6.743623   \n",
       "SIDM00078  1.443607  3.090853  4.078097  1.422233  2.333424  6.576976   \n",
       "SIDM00079  1.090853  2.157044  4.101818  2.659925  3.275007  6.530757   \n",
       "SIDM00080  0.757023  1.803227  3.493135  0.014355  2.495695  5.282440   \n",
       "SIDM00081  0.895303  1.922198  4.047015  0.963474  2.803227  5.291309   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "model_id                       \n",
       "SIDM00046  4.218781  3.653060  \n",
       "SIDM00078  3.486714  2.691534  \n",
       "SIDM00079  3.179511  3.477677  \n",
       "SIDM00080  3.845992  4.080658  \n",
       "SIDM00081  3.544733  4.852498  \n",
       "\n",
       "[5 rows x 19193 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene expression data\n",
    "exp_matrix = pd.read_csv(\n",
    "    dataset_folder / \"OmicsExpressionProteinCodingGenesTPMLogp1.csv\",\n",
    "    index_col=0,\n",
    ")\n",
    "exp_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SIDM00046</th>\n",
       "      <td>1.291901</td>\n",
       "      <td>1.291901</td>\n",
       "      <td>0.832412</td>\n",
       "      <td>1.177027</td>\n",
       "      <td>1.177027</td>\n",
       "      <td>1.177027</td>\n",
       "      <td>1.177027</td>\n",
       "      <td>0.960124</td>\n",
       "      <td>0.946083</td>\n",
       "      <td>1.220375</td>\n",
       "      <td>...</td>\n",
       "      <td>0.873702</td>\n",
       "      <td>0.832412</td>\n",
       "      <td>0.716176</td>\n",
       "      <td>0.716176</td>\n",
       "      <td>1.141600</td>\n",
       "      <td>0.960124</td>\n",
       "      <td>0.960124</td>\n",
       "      <td>1.565393</td>\n",
       "      <td>1.180616</td>\n",
       "      <td>0.960124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00078</th>\n",
       "      <td>1.764619</td>\n",
       "      <td>1.764619</td>\n",
       "      <td>0.901516</td>\n",
       "      <td>0.661356</td>\n",
       "      <td>0.661356</td>\n",
       "      <td>0.661356</td>\n",
       "      <td>0.661356</td>\n",
       "      <td>0.975263</td>\n",
       "      <td>1.093212</td>\n",
       "      <td>1.221159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.992034</td>\n",
       "      <td>0.901516</td>\n",
       "      <td>0.597052</td>\n",
       "      <td>0.597052</td>\n",
       "      <td>1.221159</td>\n",
       "      <td>0.975263</td>\n",
       "      <td>0.975263</td>\n",
       "      <td>0.974170</td>\n",
       "      <td>0.900280</td>\n",
       "      <td>0.649486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00079</th>\n",
       "      <td>1.005852</td>\n",
       "      <td>1.005852</td>\n",
       "      <td>1.064016</td>\n",
       "      <td>0.630855</td>\n",
       "      <td>0.630855</td>\n",
       "      <td>0.630855</td>\n",
       "      <td>0.630855</td>\n",
       "      <td>1.015066</td>\n",
       "      <td>1.017610</td>\n",
       "      <td>1.059105</td>\n",
       "      <td>...</td>\n",
       "      <td>1.017607</td>\n",
       "      <td>1.064016</td>\n",
       "      <td>1.034069</td>\n",
       "      <td>1.034069</td>\n",
       "      <td>1.032002</td>\n",
       "      <td>1.026443</td>\n",
       "      <td>1.026443</td>\n",
       "      <td>0.626533</td>\n",
       "      <td>0.041083</td>\n",
       "      <td>0.653868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00080</th>\n",
       "      <td>0.806532</td>\n",
       "      <td>0.806532</td>\n",
       "      <td>0.836938</td>\n",
       "      <td>1.044198</td>\n",
       "      <td>1.044198</td>\n",
       "      <td>1.044198</td>\n",
       "      <td>1.044198</td>\n",
       "      <td>0.913477</td>\n",
       "      <td>1.136541</td>\n",
       "      <td>1.118773</td>\n",
       "      <td>...</td>\n",
       "      <td>1.022343</td>\n",
       "      <td>0.901621</td>\n",
       "      <td>0.749775</td>\n",
       "      <td>0.749775</td>\n",
       "      <td>1.118773</td>\n",
       "      <td>0.913477</td>\n",
       "      <td>0.913477</td>\n",
       "      <td>1.372159</td>\n",
       "      <td>0.990136</td>\n",
       "      <td>0.997781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00081</th>\n",
       "      <td>1.206992</td>\n",
       "      <td>1.206992</td>\n",
       "      <td>0.801166</td>\n",
       "      <td>0.858750</td>\n",
       "      <td>0.858750</td>\n",
       "      <td>0.858750</td>\n",
       "      <td>0.858750</td>\n",
       "      <td>1.159133</td>\n",
       "      <td>1.325867</td>\n",
       "      <td>1.377363</td>\n",
       "      <td>...</td>\n",
       "      <td>0.858489</td>\n",
       "      <td>0.801166</td>\n",
       "      <td>0.562601</td>\n",
       "      <td>0.562601</td>\n",
       "      <td>1.377363</td>\n",
       "      <td>1.159133</td>\n",
       "      <td>1.159133</td>\n",
       "      <td>1.205385</td>\n",
       "      <td>1.018584</td>\n",
       "      <td>1.210574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25368 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A1BG  A1BG-AS1      A1CF       A2M   A2M-AS1     A2ML1  \\\n",
       "model_id                                                                \n",
       "SIDM00046  1.291901  1.291901  0.832412  1.177027  1.177027  1.177027   \n",
       "SIDM00078  1.764619  1.764619  0.901516  0.661356  0.661356  0.661356   \n",
       "SIDM00079  1.005852  1.005852  1.064016  0.630855  0.630855  0.630855   \n",
       "SIDM00080  0.806532  0.806532  0.836938  1.044198  1.044198  1.044198   \n",
       "SIDM00081  1.206992  1.206992  0.801166  0.858750  0.858750  0.858750   \n",
       "\n",
       "              A2MP1   A3GALT2    A4GALT     A4GNT  ...    ZWILCH     ZWINT  \\\n",
       "model_id                                           ...                       \n",
       "SIDM00046  1.177027  0.960124  0.946083  1.220375  ...  0.873702  0.832412   \n",
       "SIDM00078  0.661356  0.975263  1.093212  1.221159  ...  0.992034  0.901516   \n",
       "SIDM00079  0.630855  1.015066  1.017610  1.059105  ...  1.017607  1.064016   \n",
       "SIDM00080  1.044198  0.913477  1.136541  1.118773  ...  1.022343  0.901621   \n",
       "SIDM00081  0.858750  1.159133  1.325867  1.377363  ...  0.858489  0.801166   \n",
       "\n",
       "               ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "model_id                                                                \n",
       "SIDM00046  0.716176  0.716176  1.141600  0.960124  0.960124  1.565393   \n",
       "SIDM00078  0.597052  0.597052  1.221159  0.975263  0.975263  0.974170   \n",
       "SIDM00079  1.034069  1.034069  1.032002  1.026443  1.026443  0.626533   \n",
       "SIDM00080  0.749775  0.749775  1.118773  0.913477  0.913477  1.372159   \n",
       "SIDM00081  0.562601  0.562601  1.377363  1.159133  1.159133  1.205385   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "model_id                       \n",
       "SIDM00046  1.180616  0.960124  \n",
       "SIDM00078  0.900280  0.649486  \n",
       "SIDM00079  0.041083  0.653868  \n",
       "SIDM00080  0.990136  0.997781  \n",
       "SIDM00081  1.018584  1.210574  \n",
       "\n",
       "[5 rows x 25368 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load copy number data\n",
    "cnv_matrix = pd.read_csv(\n",
    "    dataset_folder / \"OmicsCNGeneCNRatioLogp1.csv\", index_col=0\n",
    ")\n",
    "\n",
    "cnv_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>AF</th>\n",
       "      <th>RefCount</th>\n",
       "      <th>AltCount</th>\n",
       "      <th>GT</th>\n",
       "      <th>PS</th>\n",
       "      <th>VariantType</th>\n",
       "      <th>...</th>\n",
       "      <th>PharmgkbID</th>\n",
       "      <th>DidaID</th>\n",
       "      <th>DidaName</th>\n",
       "      <th>GwasDisease</th>\n",
       "      <th>GwasPmID</th>\n",
       "      <th>GTexGene</th>\n",
       "      <th>ModelID</th>\n",
       "      <th>EntrezGeneID</th>\n",
       "      <th>SangerModelID</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>999959</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.556</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000943</td>\n",
       "      <td>57801.0</td>\n",
       "      <td>SIDM01090</td>\n",
       "      <td>HES4:1:999959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1046480</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>0.467</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INS</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000943</td>\n",
       "      <td>375790.0</td>\n",
       "      <td>SIDM01090</td>\n",
       "      <td>AGRN:1:1046480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1048197</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.571</td>\n",
       "      <td>21</td>\n",
       "      <td>27</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000943</td>\n",
       "      <td>375790.0</td>\n",
       "      <td>SIDM01090</td>\n",
       "      <td>AGRN:1:1048197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1196686</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.453</td>\n",
       "      <td>28</td>\n",
       "      <td>23</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000943</td>\n",
       "      <td>254173.0</td>\n",
       "      <td>SIDM01090</td>\n",
       "      <td>TTLL10:1:1196686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1313091</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.577</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH-000943</td>\n",
       "      <td>54973.0</td>\n",
       "      <td>SIDM01090</td>\n",
       "      <td>INTS11:1:1313091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chrom      Pos Ref Alt     AF  RefCount  AltCount   GT  PS VariantType  ...  \\\n",
       "0     1   999959   C   T  0.556        11        14  0/1 NaN         SNP  ...   \n",
       "1     1  1046480   G  GC  0.467        25        21  0/1 NaN         INS  ...   \n",
       "2     1  1048197   A   G  0.571        21        27  0/1 NaN         SNP  ...   \n",
       "3     1  1196686   C   A  0.453        28        23  0/1 NaN         SNP  ...   \n",
       "4     1  1313091   C   T  0.577        22        29  0/1 NaN         SNP  ...   \n",
       "\n",
       "  PharmgkbID DidaID DidaName GwasDisease GwasPmID GTexGene     ModelID  \\\n",
       "0        NaN    NaN      NaN         NaN      NaN      NaN  ACH-000943   \n",
       "1        NaN    NaN      NaN         NaN      NaN      NaN  ACH-000943   \n",
       "2        NaN    NaN      NaN         NaN      NaN      NaN  ACH-000943   \n",
       "3        NaN    NaN      NaN         NaN      NaN      NaN  ACH-000943   \n",
       "4        NaN    NaN      NaN         NaN      NaN      NaN  ACH-000943   \n",
       "\n",
       "   EntrezGeneID SangerModelID                id  \n",
       "0       57801.0     SIDM01090     HES4:1:999959  \n",
       "1      375790.0     SIDM01090    AGRN:1:1046480  \n",
       "2      375790.0     SIDM01090    AGRN:1:1048197  \n",
       "3      254173.0     SIDM01090  TTLL10:1:1196686  \n",
       "4       54973.0     SIDM01090  INTS11:1:1313091  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load mutation data\n",
    "mut_maf = pd.read_csv(\n",
    "    dataset_folder / \"OmicsSomaticMutationsHarmonizedMAF.csv\", low_memory=False\n",
    ")\n",
    "\n",
    "mut_maf[\"Chrom\"] = mut_maf[\"Chrom\"].map(lambda v: str(v).replace(\"chr\", \"\"))\n",
    "mut_maf[\"id\"] = (\n",
    "    mut_maf[[\"HugoSymbol\", \"Chrom\", \"Pos\"]].astype(str).agg(\":\".join, axis=1)\n",
    ")\n",
    "\n",
    "mut_maf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_genes = set(exp_matrix.columns)\n",
    "cnv_genes = set(cnv_matrix.columns)\n",
    "mut_genes = set(mut_maf[\"HugoSymbol\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate ScreenDL features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "screendl_input_folder = input_folder / \"ScreenDL\"\n",
    "screendl_input_folder.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "screendl_genes = set.intersection(mcg_genes, exp_genes)\n",
    "screendl_genes = sorted(list(screendl_genes))\n",
    "\n",
    "screendl_exp_matrix = exp_matrix[screendl_genes]\n",
    "\n",
    "n_genes = len(screendl_genes)\n",
    "\n",
    "# screendl_exp_matrix.to_csv(\n",
    "#     screendl_input_folder\n",
    "#     / f\"FeatureCellToExpression{n_genes}MCGGenesTPMLogp1.csv\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DeepCDR features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "deepcdr_input_folder = input_folder / \"DeepCDR\"\n",
    "deepcdr_input_folder.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "717\n"
     ]
    }
   ],
   "source": [
    "# generate expression features\n",
    "\n",
    "# deepcdr_exp_path = \"https://raw.githubusercontent.com/kimmo1019/DeepCDR/master/data/CCLE/genomic_expression_561celllines_697genes_demap_features.csv\"\n",
    "# deepcdr_exp_genes = pd.read_csv(\n",
    "#     deepcdr_exp_path, index_col=0, nrows=1\n",
    "# ).columns.to_list()\n",
    "# deepcdr_exp_genes = set(deepcdr_exp_genes)\n",
    "deepcdr_exp_genes = cgc_genes\n",
    "\n",
    "deepcdr_exp_genes_filtered = set.intersection(deepcdr_exp_genes, exp_genes)\n",
    "deepcdr_exp_genes_filtered = sorted(list(deepcdr_exp_genes_filtered))\n",
    "print(len(deepcdr_exp_genes_filtered))\n",
    "\n",
    "deepcdr_exp_matrix = exp_matrix[deepcdr_exp_genes_filtered]\n",
    "n_genes = len(deepcdr_exp_genes_filtered)\n",
    "\n",
    "# deepcdr_exp_matrix.to_csv(\n",
    "#     deepcdr_input_folder\n",
    "#     / f\"FeatureCellToExpression{n_genes}CGCGenesTPMLogp1.csv\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "716\n"
     ]
    }
   ],
   "source": [
    "# generate mutation features\n",
    "\n",
    "# deepcdr_mut_path = \"https://raw.githubusercontent.com/kimmo1019/DeepCDR/master/data/CCLE/genomic_mutation_34673_demap_features.csv\"\n",
    "# deepcdr_mut_genes = (\n",
    "#     pd.read_csv(deepcdr_mut_path, index_col=0, nrows=2)\n",
    "#     .columns.map(lambda c: c.split(\".\")[0])\n",
    "#     .unique()\n",
    "#     .tolist()\n",
    "# )\n",
    "# deepcdr_mut_genes = set(deepcdr_mut_genes)\n",
    "deepcdr_mut_genes = cgc_genes\n",
    "deepcdr_mut_genes_filtered = set.intersection(deepcdr_mut_genes, mut_genes)\n",
    "print(len(deepcdr_mut_genes_filtered))\n",
    "\n",
    "deepcdr_mut_maf = mut_maf[mut_maf[\"HugoSymbol\"].isin(deepcdr_mut_genes)]\n",
    "\n",
    "# lazy check to make sure we didn't lose any cell lines\n",
    "old_cells = mut_maf[\"SangerModelID\"].unique().tolist()\n",
    "new_cells = deepcdr_mut_maf[\"SangerModelID\"].unique().tolist()\n",
    "assert set(old_cells) == set(new_cells)\n",
    "\n",
    "deepcdr_mut_matrix = (\n",
    "    deepcdr_mut_maf.groupby([\"SangerModelID\", \"id\"])\n",
    "    .size()\n",
    "    .unstack()\n",
    "    .fillna(0)\n",
    "    .clip(upper=1)\n",
    "    .astype(int)\n",
    ")\n",
    "n_genes = len(deepcdr_mut_genes_filtered)\n",
    "\n",
    "positions = deepcdr_mut_matrix.columns.to_list()\n",
    "positions = pd.DataFrame({\"pos\": positions})\n",
    "positions[\"gene\"] = positions[\"pos\"].map(lambda v: str(v).split(\":\")[0])\n",
    "positions[\"chr\"] = positions[\"pos\"].map(lambda v: str(v).split(\":\")[1])\n",
    "positions[\"start\"] = positions[\"pos\"].map(lambda v: int(str(v).split(\":\")[2]))\n",
    "\n",
    "mapping = {str(i): i for i in range(1, 23)}\n",
    "mapping[\"X\"] = 23\n",
    "mapping[\"Y\"] = 24\n",
    "positions[\"chr\"] = positions[\"chr\"].map(mapping)\n",
    "\n",
    "sorted_positions = positions.sort_values([\"chr\", \"start\"])[\"pos\"].to_list()\n",
    "\n",
    "deepcdr_mut_matrix = deepcdr_mut_matrix[sorted_positions]\n",
    "\n",
    "deepcdr_mut_matrix.to_csv(\n",
    "    deepcdr_input_folder\n",
    "    / f\"FeatureCellToSomaticMutationsPositionEncoded{n_genes}CGCGenesAll.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate DualGCN features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dualgcn_input_folder = input_folder / \"DualGCN\"\n",
    "dualgcn_input_folder.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HOXA11</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HOXA11</td>\n",
       "      <td>JUN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_1  gene_2\n",
       "0  HOXA11  CDKN2A\n",
       "1  HOXA11     JUN"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dualgcn_ppi = pd.read_csv(\n",
    "    dualgcn_folder / \"PPI_network.txt\",\n",
    "    sep=\"\\t\",\n",
    "    usecols=[0, 1],\n",
    "    names=[\"gene_1\", \"gene_2\"],\n",
    "    header=None,\n",
    ")\n",
    "\n",
    "dualgcn_ppi_genes = set.intersection(\n",
    "    set(dualgcn_ppi[\"gene_1\"].to_list()),\n",
    "    set(dualgcn_ppi[\"gene_2\"].to_list()),\n",
    ")\n",
    "\n",
    "dualgcn_ppi.head(n=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dualgcn_genes_filtered = set.intersection(\n",
    "    dualgcn_genes, dualgcn_ppi_genes, exp_genes, cnv_genes\n",
    ")\n",
    "dualgcn_genes_filtered = sorted(list(dualgcn_genes_filtered))\n",
    "\n",
    "dualgcn_exp_matrix = exp_matrix[dualgcn_genes_filtered]\n",
    "dualgcn_cnv_matrix = cnv_matrix[dualgcn_genes_filtered]\n",
    "dualgcn_ppi = dualgcn_ppi[\n",
    "    (dualgcn_ppi[\"gene_1\"].isin(dualgcn_genes_filtered))\n",
    "    & (dualgcn_ppi[\"gene_2\"].isin(dualgcn_genes_filtered))\n",
    "]\n",
    "\n",
    "n_genes = len(dualgcn_genes_filtered)\n",
    "\n",
    "# dualgcn_exp_matrix.to_csv(\n",
    "#     dualgcn_input_folder\n",
    "#     / f\"FeatureCellToExpression{n_genes}DualGCNGenesTPMLogp1.csv\"\n",
    "# )\n",
    "\n",
    "# dualgcn_cnv_matrix.to_csv(\n",
    "#     dualgcn_input_folder\n",
    "#     / f\"FeatureCellToCopyNumber{n_genes}DualGCNGenesCNRatioLogp1.csv\"\n",
    "# )\n",
    "\n",
    "# dualgcn_ppi.to_csv(\n",
    "#     dualgcn_input_folder / \"MetadataPPIEdgeList.csv\",\n",
    "#     index=False,\n",
    "# )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate mutation features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load mutation data\n",
    "mut_data_maf = pd.read_csv(\n",
    "    dataset_folder / \"OmicsSomaticMutationsHarmonizedMAF.csv\",\n",
    "    low_memory=False,\n",
    ")\n",
    "\n",
    "mut_data_maf[\"Chrom\"] = mut_data_maf[\"Chrom\"].map(\n",
    "    lambda v: str(v).replace(\"chr\", \"\")\n",
    ")\n",
    "mut_data_maf[\"id\"] = (\n",
    "    mut_data_maf[[\"HugoSymbol\", \"Chrom\", \"Pos\"]]\n",
    "    .astype(str)\n",
    "    .agg(\":\".join, axis=1)\n",
    ")\n",
    "\n",
    "mut_data_maf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# positional encoding of mutations for convolutional models\n",
    "mut_data_cgc_genes = mut_data_maf[\n",
    "    mut_data_maf[\"HugoSymbol\"].isin(cgc_genes)\n",
    "]\n",
    "\n",
    "# lazy check to make sure we didn't lose any cell lines\n",
    "old_cell_ids = mut_data_maf[\"SangerModelID\"].unique().tolist()\n",
    "new_cell_ids = mut_data_cgc_genes[\"SangerModelID\"].unique().tolist()\n",
    "assert set(old_cell_ids) == set(new_cell_ids)\n",
    "\n",
    "mut_matrix = (\n",
    "    mut_data_cgc_genes.groupby([\"SangerModelID\", \"id\"])\n",
    "    .size()\n",
    "    .unstack()\n",
    "    .fillna(0)\n",
    "    .clip(upper=1)\n",
    "    .astype(int)\n",
    ")\n",
    "\n",
    "n_genes = mut_data_cgc_genes[\"HugoSymbol\"].nunique()\n",
    "\n",
    "# mut_matrix.to_csv(\n",
    "#     input_folder\n",
    "#     / f\"CellToSomaticMutationsPositionEncoded{n_genes}CGCGenesAll.csv\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all mutations encoded by gene\n",
    "mut_matrix_all = (\n",
    "    mut_data_maf.groupby([\"SangerModelID\", \"HugoSymbol\"])\n",
    "    .size()\n",
    "    .unstack()\n",
    "    .fillna(0)\n",
    "    .clip(upper=1)\n",
    "    .astype(int)\n",
    ")\n",
    "\n",
    "mut_matrix_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene encoded matrix CGC genes\n",
    "mut_matrix_all_cgc = mut_matrix_all.filter(items=cgc_genes, axis=1)\n",
    "mut_matrix_all_cgc = mut_matrix_all_cgc.sort_index(axis=1)\n",
    "\n",
    "n_genes = mut_matrix_all_cgc.shape[1]\n",
    "\n",
    "# mut_matrix_all_cgc.to_csv(\n",
    "#     input_folder / f\"CellToSomaticMutationsGeneEncoded{n_genes}CGCGenesAll.csv\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_matrix_del = (\n",
    "    mut_data_maf[mut_data_maf[\"CCLEDeleterious\"] == True]\n",
    "    .groupby([\"SangerModelID\", \"HugoSymbol\"])\n",
    "    .size()\n",
    "    .unstack()\n",
    "    .fillna(0)\n",
    "    .clip(upper=1)\n",
    "    .astype(int)\n",
    ")\n",
    "\n",
    "mut_matrix_del"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mut_data_cgc_genes[\n",
    "#     (mut_data_cgc_genes[\"CCLEDeleterious\"] == True)\n",
    "#     | (mut_data_cgc_genes[\"HessDriver\"] == True)\n",
    "# ].shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cdrpy-tf-cpu",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
