{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-18 18:59:40.150639: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-07-18 18:59:41.816012: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "from __future__ import annotations\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tensorflow as tf\n",
    "import typing as t\n",
    "\n",
    "from pathlib import Path\n",
    "from tensorflow import keras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cdrpy.feat.encoders import PandasEncoder, DictEncoder\n",
    "from cdrpy.models.dualgcn.util import normalize_adj\n",
    "from cdrpy.util.types import PathLike\n",
    "from cdrpy.util.io import read_pickled_dict\n",
    "\n",
    "DegList = list[np.int32]\n",
    "AdjList = list[list[int]]\n",
    "ConvMolFeat = tuple[np.ndarray, DegList, AdjList]\n",
    "\n",
    "\n",
    "def _process_drug_feature(\n",
    "    feat_mat: np.ndarray, adj_lst: AdjList, max_atoms: int = 100\n",
    ") -> tuple[np.ndarray, np.ndarray]:\n",
    "    \"\"\"Add padding and convert to `np.ndarray`.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "        feat_mat:\n",
    "        adj_lst:\n",
    "        max_atoms:\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        A tuple of (feature_matrix, adjacency_matrix).\n",
    "    \"\"\"\n",
    "    n_atoms = len(adj_lst)\n",
    "    assert feat_mat.shape[0] == n_atoms\n",
    "\n",
    "    F = np.zeros((max_atoms, feat_mat.shape[-1]))\n",
    "    F[: feat_mat.shape[0], :] = feat_mat\n",
    "    A = np.zeros((max_atoms, max_atoms), dtype=\"float32\")\n",
    "\n",
    "    for i, nodes in enumerate(adj_lst):\n",
    "        for n in nodes:\n",
    "            A[i, int(n)] = 1\n",
    "\n",
    "    assert np.allclose(A, A.T)\n",
    "\n",
    "    A_vals = normalize_adj(A[:n_atoms, :n_atoms])\n",
    "    A_pad = normalize_adj(A[n_atoms:, n_atoms:])\n",
    "\n",
    "    A[:n_atoms, :n_atoms] = A_vals\n",
    "    A[n_atoms:, n_atoms:] = A_pad\n",
    "\n",
    "    return F, A\n",
    "\n",
    "\n",
    "def load_drug_features(\n",
    "    mol_path: PathLike | Path,\n",
    ") -> tuple[DictEncoder, DictEncoder]:\n",
    "    \"\"\"Load drug chemical features.\"\"\"\n",
    "    drug_dict = read_pickled_dict(mol_path)\n",
    "    drug_dict = {\n",
    "        k: _process_drug_feature(F, A) for k, (F, _, A) in drug_dict.items()\n",
    "    }\n",
    "    drug_feat = {k: v[0].astype(\"float32\") for k, v in drug_dict.items()}\n",
    "    drug_adj = {k: v[1].astype(\"float32\") for k, v in drug_dict.items()}\n",
    "    return (\n",
    "        DictEncoder(drug_feat, name=\"drug_feature_encoder\"),\n",
    "        DictEncoder(drug_adj, name=\"drug_adj_encoder\"),\n",
    "    )\n",
    "\n",
    "\n",
    "def load_cell_features(\n",
    "    exp_path: PathLike | Path,\n",
    "    mut_path: PathLike | Path,\n",
    "    methyl_path: PathLike | Path | None = None,\n",
    ") -> tuple[PandasEncoder, DictEncoder, PandasEncoder | None]:\n",
    "    \"\"\"Load cell features for ScreenDL.\"\"\"\n",
    "    exp_mat = pd.read_csv(exp_path, index_col=0).astype(\"float32\")\n",
    "    mut_mat = pd.read_csv(mut_path, index_col=0).astype(\"int32\")\n",
    "    \n",
    "    mut_dict = {}\n",
    "    for cell_id in mut_mat.index:\n",
    "        cell_mut = mut_mat.loc[cell_id].values.reshape(1, -1, 1)\n",
    "        mut_dict[cell_id] = cell_mut\n",
    "\n",
    "    methyl_enc = None\n",
    "    if methyl_path is not None:\n",
    "        methyl_mat = pd.read_csv(methyl_path, index_col=0).astype(\"float32\")\n",
    "        methyl_enc = PandasEncoder(methyl_mat, name=\"methyl_encoder\")\n",
    "\n",
    "    return (\n",
    "        PandasEncoder(exp_mat, name=\"exp_encoder\"),\n",
    "        DictEncoder(mut_dict, name=\"mut_encoder\"),\n",
    "        methyl_enc,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = Path(\"../../data/inputs/GDSCv2DepMap\")\n",
    "exp_path = input_dir / \"DeepCDR/FeatureCellToExpression717CGCGenesTPMLogp1.csv\"\n",
    "mut_path = input_dir / \"DeepCDR/FeatureCellToSomaticMutationsPositionEncoded716CGCGenesAll.csv\"\n",
    "mol_path = input_dir / \"DrugToConvMolFeatures.pickle\"\n",
    "\n",
    "cell_exp_enc, cell_mut_enc, _ = load_cell_features(exp_path, mut_path)\n",
    "drug_feat_enc, drug_adj_enc = load_drug_features(mol_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_mat = pd.read_csv(mut_path, index_col=0, nrows=10).astype(\"int32\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import keras.backend as K\n",
    "\n",
    "from keras import layers\n",
    "from cdrpy.layers.graph import GraphConvBlock\n",
    "\n",
    "\n",
    "def _create_mut_subnetwork(mut_dim: int) -> keras.Sequential:\n",
    "    \"\"\"\"\"\"\n",
    "    model = keras.Sequential(\n",
    "        [\n",
    "            layers.Input((1, mut_dim, 1), name=\"cell_mut_input\"),\n",
    "            layers.Conv2D(\n",
    "                50,\n",
    "                activation=\"tanh\",\n",
    "                kernel_size=(1, 700),\n",
    "                strides=(1, 5),\n",
    "                name=\"cell_mut_conv2d_1\",\n",
    "            ),\n",
    "            layers.MaxPool2D(pool_size=(1, 5), name=\"cell_mut_mpool2d_1\"),\n",
    "            layers.Conv2D(\n",
    "                30,\n",
    "                activation=\"relu\",\n",
    "                kernel_size=(1, 5),\n",
    "                strides=(1, 2),\n",
    "                name=\"cell_mut_conv2d_2\",\n",
    "            ),\n",
    "            layers.MaxPooling2D(pool_size=(1, 10), name=\"cell_mut_mpool2d_2\"),\n",
    "            layers.Flatten(name=\"cell_mut_flatten_1\"),\n",
    "            layers.Dense(100, activation=\"relu\", name=\"cell_mut_dense_1\"),\n",
    "            layers.Dropout(0.1, name=\"cell_mut_dropout_1\"),\n",
    "        ],\n",
    "        name=\"cell_mut_subnet\",\n",
    "    )\n",
    "\n",
    "    return model\n",
    "\n",
    "\n",
    "def _create_exp_subnetwork(\n",
    "    exp_dim: int, exp_norm: layers.Normalization | None = None\n",
    ") -> keras.Model:\n",
    "    \"\"\"Creates the gene expression subnetwork.\"\"\"\n",
    "    exp_input = layers.Input((exp_dim,), name=\"cell_exp_input\")\n",
    "    if exp_norm is not None:\n",
    "        if not exp_norm.is_adapted:\n",
    "            # FIXME: change this to a warning since you can still adapt later\n",
    "            raise ValueError(\"requires adapted normalization layer...\")\n",
    "        exp_input = exp_norm(exp_input)\n",
    "    x = layers.Dense(256, activation=\"tanh\", name=\"cell_exp_dense_1\")(exp_input)\n",
    "    x = layers.BatchNormalization(name=\"cell_exp_bnorm_1\")(x)\n",
    "    x = layers.Dropout(0.1, name=\"cell_exp_dropout_1\")(x)\n",
    "    x = layers.Dense(100, activation=\"relu\", name=\"cell_exp_dense_2\")(x)\n",
    "\n",
    "    model = keras.Model(inputs=exp_input, outputs=x, name=\"cell_exp_subnet\")\n",
    "\n",
    "    return model\n",
    "\n",
    "\n",
    "def _create_methyl_subnetwork(methyl_dim: int) -> keras.Sequential:\n",
    "    \"\"\"Creates the methylation subnetwork.\"\"\"\n",
    "    # FIXME: add option for normalization layer since this is float data\n",
    "    model = keras.Sequential(\n",
    "        [\n",
    "            layers.Input((methyl_dim,), name=\"cell_methyl_input\"),\n",
    "            layers.Dense(256, activation=\"tanh\", name=\"cell_methyl_dense_1\"),\n",
    "            layers.BatchNormalization(name=\"cell_methyl_bnorm_1\"),\n",
    "            layers.Dropout(0.1, name=\"cell_methyl_dropout_1\"),\n",
    "            layers.Dense(100, activation=\"relu\", name=\"cell_methyl_dense_2\"),\n",
    "        ],\n",
    "        name=\"cell_methyl_subnet\",\n",
    "    )\n",
    "\n",
    "\n",
    "def _create_cell_subnetwork(\n",
    "    exp_dim: int,\n",
    "    mut_dim: int,\n",
    "    methyl_dim: int | None = None,\n",
    "    exp_norm: layers.Normalization | None = None,\n",
    ") -> keras.Model:\n",
    "    \"\"\"\"\"\"\n",
    "    exp_subnet = _create_exp_subnetwork(exp_dim, exp_norm)\n",
    "    mut_subnet = _create_mut_subnetwork(mut_dim)\n",
    "    subnet_inputs = [exp_subnet.input, mut_subnet.input]\n",
    "    subnet_outputs = [exp_subnet.output, mut_subnet.output]\n",
    "    if methyl_dim is not None:\n",
    "        methyl_subnet = _create_methyl_subnetwork(methyl_dim)\n",
    "        subnet_inputs.append(methyl_subnet.input)\n",
    "        subnet_outputs.append(methyl_subnet.output)\n",
    "    output = layers.Concatenate(name=\"cell_subnet_concat\")(subnet_outputs)\n",
    "\n",
    "    return keras.Model(\n",
    "        inputs=subnet_inputs, outputs=output, name=\"cell_subnet\"\n",
    "    )\n",
    "\n",
    "\n",
    "def _create_drug_subnetwork(feat_dim: int) -> keras.Model:\n",
    "    \"\"\"Creates the drug subnetwork.\"\"\"\n",
    "    feat_input = layers.Input((None, feat_dim), name=\"drug_feat_input\")\n",
    "    adj_input = layers.Input((None, None), name=\"drug_adj_input\")\n",
    "\n",
    "    x = [feat_input, adj_input]\n",
    "    x = GraphConvBlock(units=256, step_num=1, name=\"drug_gconv_1\")(x)\n",
    "    x = [x, adj_input]\n",
    "    x = GraphConvBlock(units=256, step_num=1, name=\"drug_gconv_2\")(x)\n",
    "    x = [x, adj_input]\n",
    "    x = GraphConvBlock(units=256, step_num=1, name=\"drug_gconv_3\")(x)\n",
    "    x = [x, adj_input]\n",
    "    x = GraphConvBlock(units=100, step_num=1, name=\"drug_gconv_4\")(x)\n",
    "    x = layers.GlobalMaxPooling1D(name=\"drug_pool\")(x)\n",
    "\n",
    "    return keras.Model(\n",
    "        inputs=[feat_input, adj_input], outputs=x, name=\"drug_subnet\"\n",
    "    )\n",
    "\n",
    "\n",
    "def create_model(\n",
    "    cell_exp_dim: int,\n",
    "    cell_mut_dim: int,\n",
    "    drug_feat_dim: int,\n",
    "    cell_methyl_dim: int | None = None,\n",
    "    cell_exp_norm: layers.Normalization | None = None,\n",
    ") -> keras.Model:\n",
    "    \"\"\"Creates the DeepCDR model.\"\"\"\n",
    "    cell_subnet = _create_cell_subnetwork(\n",
    "        cell_exp_dim, cell_mut_dim, cell_methyl_dim, cell_exp_norm\n",
    "    )\n",
    "    drug_subnet = _create_drug_subnetwork(drug_feat_dim)\n",
    "\n",
    "    subnet_inputs = [*cell_subnet.input, *drug_subnet.input]\n",
    "    subnet_outputs = [cell_subnet.output, drug_subnet.output]\n",
    "\n",
    "    x = layers.Concatenate(name=\"shared_concat_1\")(subnet_outputs)\n",
    "    x = layers.Dense(300, activation=\"tanh\", name=\"shared_dense_1\")(x)\n",
    "    x = layers.Dropout(0.1, name=\"shared_dropout_1\")(x)\n",
    "    x = layers.Lambda(lambda x: K.expand_dims(x, axis=-1))(x)\n",
    "    x = layers.Lambda(lambda x: K.expand_dims(x, axis=1))(x)\n",
    "    x = layers.Conv2D(\n",
    "        30,\n",
    "        kernel_size=(1, 150),\n",
    "        strides=(1, 1),\n",
    "        activation=\"relu\",\n",
    "        name=\"shared_conv2d_1\",\n",
    "    )(x)\n",
    "    x = layers.MaxPooling2D(pool_size=(1, 2), name=\"shared_mpool2d_1\")(x)\n",
    "    x = layers.Conv2D(\n",
    "        filters=10,\n",
    "        kernel_size=(1, 5),\n",
    "        strides=(1, 1),\n",
    "        activation=\"relu\",\n",
    "        name=\"shared_conv2d_2\",\n",
    "    )(x)\n",
    "    x = layers.MaxPooling2D(pool_size=(1, 3), name=\"shared_mpool2d_2\")(x)\n",
    "    x = layers.Conv2D(\n",
    "        filters=5,\n",
    "        kernel_size=(1, 5),\n",
    "        strides=(1, 1),\n",
    "        activation=\"relu\",\n",
    "        name=\"shared_conv2d_3\",\n",
    "    )(x)\n",
    "    x = layers.MaxPooling2D(pool_size=(1, 3), name=\"shared_mpool2d_3\")(x)\n",
    "    x = layers.Dropout(0.1, name=\"shared_dropout_2\")(x)\n",
    "    x = layers.Flatten(name=\"shared_flat_1\")(x)\n",
    "    x = layers.Dropout(0.2, name=\"shared_dropout_3\")(x)\n",
    "    output = layers.Dense(1, name=\"output\")(x)\n",
    "\n",
    "    model = keras.Model(inputs=subnet_inputs, outputs=output, name=\"DeepCDR\")\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapt_data = np.array(\n",
    "    [[0.0, 7.0, 4.0], [2.0, 9.0, 6.0], [0.0, 7.0, 4.0], [2.0, 9.0, 6.0]],\n",
    "    dtype=\"float32\",\n",
    ")\n",
    "input_data = np.array([[1.0, 2.0, 3.0]], dtype=\"float32\")\n",
    "layer = layers.Normalization(axis=-1, invert=True)\n",
    "layer.adapt(adapt_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/uufs/chpc.utah.edu/common/HIPAA/u0871891/software/pkg/miniconda3/envs/cdrpy-tf-cpu-v2/lib/python3.9/site-packages/cdrpy/util/decorators.py:21: UserWarning: DictEncoder.shape may result in unexpected behavior.\n",
      "  warnings.warn(*warning_args, **warning_kwargs)\n"
     ]
    }
   ],
   "source": [
    "exp_dim = cell_exp_enc.shape[-1]\n",
    "mut_dim = cell_mut_enc.shape[1]\n",
    "drug_dim = drug_feat_enc.shape[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = create_model(exp_dim, mut_dim, drug_dim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"DeepCDR\"\n",
      "__________________________________________________________________________________________________\n",
      " Layer (type)                Output Shape                 Param #   Connected to                  \n",
      "==================================================================================================\n",
      " cell_mut_input (InputLayer  [(None, 1, 33290, 1)]        0         []                            \n",
      " )                                                                                                \n",
      "                                                                                                  \n",
      " cell_mut_conv2d_1 (Conv2D)  (None, 1, 6519, 50)          35050     ['cell_mut_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_mut_mpool2d_1 (MaxPoo  (None, 1, 1303, 50)          0         ['cell_mut_conv2d_1[0][0]']   \n",
      " ling2D)                                                                                          \n",
      "                                                                                                  \n",
      " cell_exp_input (InputLayer  [(None, 717)]                0         []                            \n",
      " )                                                                                                \n",
      "                                                                                                  \n",
      " cell_mut_conv2d_2 (Conv2D)  (None, 1, 650, 30)           7530      ['cell_mut_mpool2d_1[0][0]']  \n",
      "                                                                                                  \n",
      " drug_feat_input (InputLaye  [(None, None, 75)]           0         []                            \n",
      " r)                                                                                               \n",
      "                                                                                                  \n",
      " drug_adj_input (InputLayer  [(None, None, None)]         0         []                            \n",
      " )                                                                                                \n",
      "                                                                                                  \n",
      " cell_exp_dense_1 (Dense)    (None, 256)                  183808    ['cell_exp_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_mut_mpool2d_2 (MaxPoo  (None, 1, 65, 30)            0         ['cell_mut_conv2d_2[0][0]']   \n",
      " ling2D)                                                                                          \n",
      "                                                                                                  \n",
      " drug_gconv_1 (GraphConvBlo  (None, None, 256)            20480     ['drug_feat_input[0][0]',     \n",
      " ck)                                                                 'drug_adj_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_exp_bnorm_1 (BatchNor  (None, 256)                  1024      ['cell_exp_dense_1[0][0]']    \n",
      " malization)                                                                                      \n",
      "                                                                                                  \n",
      " cell_mut_flatten_1 (Flatte  (None, 1950)                 0         ['cell_mut_mpool2d_2[0][0]']  \n",
      " n)                                                                                               \n",
      "                                                                                                  \n",
      " drug_gconv_2 (GraphConvBlo  (None, None, 256)            66816     ['drug_gconv_1[0][0]',        \n",
      " ck)                                                                 'drug_adj_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_exp_dropout_1 (Dropou  (None, 256)                  0         ['cell_exp_bnorm_1[0][0]']    \n",
      " t)                                                                                               \n",
      "                                                                                                  \n",
      " cell_mut_dense_1 (Dense)    (None, 100)                  195100    ['cell_mut_flatten_1[0][0]']  \n",
      "                                                                                                  \n",
      " drug_gconv_3 (GraphConvBlo  (None, None, 256)            66816     ['drug_gconv_2[0][0]',        \n",
      " ck)                                                                 'drug_adj_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_exp_dense_2 (Dense)    (None, 100)                  25700     ['cell_exp_dropout_1[0][0]']  \n",
      "                                                                                                  \n",
      " cell_mut_dropout_1 (Dropou  (None, 100)                  0         ['cell_mut_dense_1[0][0]']    \n",
      " t)                                                                                               \n",
      "                                                                                                  \n",
      " drug_gconv_4 (GraphConvBlo  (None, None, 100)            26100     ['drug_gconv_3[0][0]',        \n",
      " ck)                                                                 'drug_adj_input[0][0]']      \n",
      "                                                                                                  \n",
      " cell_subnet_concat (Concat  (None, 200)                  0         ['cell_exp_dense_2[0][0]',    \n",
      " enate)                                                              'cell_mut_dropout_1[0][0]']  \n",
      "                                                                                                  \n",
      " drug_pool (GlobalMaxPoolin  (None, 100)                  0         ['drug_gconv_4[0][0]']        \n",
      " g1D)                                                                                             \n",
      "                                                                                                  \n",
      " shared_concat_1 (Concatena  (None, 300)                  0         ['cell_subnet_concat[0][0]',  \n",
      " te)                                                                 'drug_pool[0][0]']           \n",
      "                                                                                                  \n",
      " shared_dense_1 (Dense)      (None, 300)                  90300     ['shared_concat_1[0][0]']     \n",
      "                                                                                                  \n",
      " shared_dropout_1 (Dropout)  (None, 300)                  0         ['shared_dense_1[0][0]']      \n",
      "                                                                                                  \n",
      " lambda (Lambda)             (None, 300, 1)               0         ['shared_dropout_1[0][0]']    \n",
      "                                                                                                  \n",
      " lambda_1 (Lambda)           (None, 1, 300, 1)            0         ['lambda[0][0]']              \n",
      "                                                                                                  \n",
      " shared_conv2d_1 (Conv2D)    (None, 1, 151, 30)           4530      ['lambda_1[0][0]']            \n",
      "                                                                                                  \n",
      " shared_mpool2d_1 (MaxPooli  (None, 1, 75, 30)            0         ['shared_conv2d_1[0][0]']     \n",
      " ng2D)                                                                                            \n",
      "                                                                                                  \n",
      " shared_conv2d_2 (Conv2D)    (None, 1, 71, 10)            1510      ['shared_mpool2d_1[0][0]']    \n",
      "                                                                                                  \n",
      " shared_mpool2d_2 (MaxPooli  (None, 1, 23, 10)            0         ['shared_conv2d_2[0][0]']     \n",
      " ng2D)                                                                                            \n",
      "                                                                                                  \n",
      " shared_conv2d_3 (Conv2D)    (None, 1, 19, 5)             255       ['shared_mpool2d_2[0][0]']    \n",
      "                                                                                                  \n",
      " shared_mpool2d_3 (MaxPooli  (None, 1, 6, 5)              0         ['shared_conv2d_3[0][0]']     \n",
      " ng2D)                                                                                            \n",
      "                                                                                                  \n",
      " shared_dropout_2 (Dropout)  (None, 1, 6, 5)              0         ['shared_mpool2d_3[0][0]']    \n",
      "                                                                                                  \n",
      " shared_flat_1 (Flatten)     (None, 30)                   0         ['shared_dropout_2[0][0]']    \n",
      "                                                                                                  \n",
      " shared_dropout_3 (Dropout)  (None, 30)                   0         ['shared_flat_1[0][0]']       \n",
      "                                                                                                  \n",
      " output (Dense)              (None, 1)                    31        ['shared_dropout_3[0][0]']    \n",
      "                                                                                                  \n",
      "==================================================================================================\n",
      "Total params: 725050 (2.77 MB)\n",
      "Trainable params: 722802 (2.76 MB)\n",
      "Non-trainable params: 2248 (8.78 KB)\n",
      "__________________________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<KerasTensor: shape=(None, 200) dtype=float32 (created by layer 'cell_subnet_concat')>,\n",
       " <KerasTensor: shape=(None, 100) dtype=float32 (created by layer 'drug_pool')>]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.], dtype=float32)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.zeros((800, 1, 100, 1), dtype=\"float32\")[0][0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cdrpy-tf-cpu-v2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
